include/eosio/vm/vector.hpp

```diff
@@ -4,8 +4,70 @@
 
 #include <utility>
 #include <string>
+#include <vector>
 
 namespace eosio { namespace vm {
+   template <typename T>
+   class unmanaged_vector {
+      public:
+         constexpr unmanaged_vector(size_t size=0) : _data(size) {}
+         constexpr unmanaged_vector(const unmanaged_vector& v) = delete;
+         constexpr unmanaged_vector(unmanaged_vector&& v) : _data(std::move(v._data)), _index(v._index) {}
+
+         constexpr unmanaged_vector& operator=(unmanaged_vector&& v) {
+            _data = std::move(v._data);
+            _index = v._index;
+            return *this;
+         }
+
+         constexpr inline void resize( size_t size ) {
+            _data.resize(size);
+         }
+
+         constexpr inline void push_back( const T& val ) {
+            if ( _index >= _data.size() )
+               resize( _data.size() ? _data.size() * 2 : 1 );
+            _data[_index++] = val;
+         }
+
+         constexpr inline void emplace_back( T&& val ) {
+            if ( _index >= _data.size() )
+               resize( _data.size() ? _data.size() * 2 : 1 );
+            _data[_index++] = std::move(val);
+         }
+
+         constexpr inline void back() {
+            return _data[_index];
+         }
+
+         constexpr inline void pop_back() {
+            EOS_VM_ASSERT( _index-- > 0, wasm_vector_oob_exception, "vector pop out of bounds" );
+         }
+
+         constexpr inline T& at( size_t i ) {
+            // TODO: fix this to be more robust, assuming that the read 'i' will be in close to the max index
+            // this is the usage pattern currently by eos-vm, a better solution will be created for the abstraction
+            // of the various points of validation
```

Comment:
Does eos-vm ever access after the current end?  The fact the this has different behavior from the const version is scary.